### PR TITLE
Reland "[NVPTX] Add support for atomic add for f16 type"

### PR DESCRIPTION
**Pull request by akuegel**
atom.add.noftz.f16 is supported since SM 7.0

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td [hunk 1]
```diff
@@ -1520,11 +1520,13 @@ multiclass F_ATOMIC_2_imp<ValueType ptrT, NVPTXRegClass ptrclass,
   def imm : NVPTXInst<(outs regclass:$dst), (ins ptrclass:$addr, IMMType:$b),
     !strconcat("atom", SpaceStr, OpcStr, TypeStr, " \t$dst, [$addr], $b;", ""),
     [(set (regT regclass:$dst), (IntOp (ptrT ptrclass:$addr), IMM:$b))]>,
-  Requires<Pred>;
+  Requires<!listconcat(Pred, [!if(!eq(TypeStr, ".f16"), Predicate<"false">, Predicate<"true">)])>;
```
Comment (akuegel) [inline llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1523]: This is a bit ugly, maybe you have an idea how to improve this? At least it seems to do what it should :)

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td [hunk 2]
```diff
@@ -1520,11 +1520,13 @@ multiclass F_ATOMIC_2_imp<ValueType ptrT, NVPTXRegClass ptrclass,
...
 }
 multiclass F_ATOMIC_2<ValueType regT, NVPTXRegClass regclass, string SpaceStr, string TypeStr,
   string OpcStr, PatFrag IntOp, Operand IMMType, SDNode IMM,
   list<Predicate> Pred = []> {
+  defm p16 : F_ATOMIC_2_imp<i16, Int16Regs, regT, regclass, SpaceStr, TypeStr, OpcStr,
```
Comment (akuegel) [inline llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1528]: When debugging, I thought that we may need this (pattern matching what exists), but not 100% sure as I don't understand what it does.

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td (hunk 1 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1523]: Considering that we want to disable `imm` completely, a somewhat simpler way would be to do
```
Requires<!if(!eq(TypeStr, ".f16"), [Predicate<"false">], Preds)])
```

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td (hunk 2 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1528]: You do not need this. p32/p64 is about pointer size and is not affected by adding an fp16 data type.

**Review by Artem-B — APPROVED** [review]
LGTM

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td (hunk 1 above)
Comment (akuegel) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td]: Thanks for the suggestion, this looks much nicer :)
Done.

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td (hunk 2 above)
Comment (akuegel) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td]: Thanks for the explanation. Removed it again.

**Pull request merged**